### DISCUSSION TITLE
Updating models?

### DISCUSSION THREAD

**Discussion by igr** [vexy/Fridge#51] [Q&A] [answered]
What happens when a model gets updated?

It seems that Fridge fails to unfreeze existing data into a new structure. I see two cases here:

1. _extended_ model (e.g., new field added): I expected this to work when a new field has a default value.
2. _modified_ model (e.g., existing fields changes type etc): this is ok to fail.

One idea is to catch errors on load and then try to _import_ data by first loading them into a generic dictionary/map (whatever is available on swift) and then manually importing data into the new, extended structure. This process could also give a mapping callback that could be used in case of modified fields (e.g. `(fieldName, oldValue) -> newValue`)

**Comment by vexy** [comment, marked as answer]
> What happens when a model gets updated?

Unfortunately, Fridge fails to unfreeze a model that gets changed meanwhile. Reason for this unwanted behavior is internals of the _freezing_ sub-system (file storage).

As this may lead to extra responsibility of the storage layer and increase complexity - which is tried to be avoided by "moving" these tasks to a upper layers - this is known limitation.

However, this may be sort of _workaround_ solution:
- `drop` original model
- setup new (changed structure)
- `freeze` it again

Eventually, this may be wrapped into standalone function, named eg. `swap(..)` ?

**Reply by igr** [reply]
That's the thing, you can't drop an existing model as it contains production data.

**Reply by vexy** [reply]
True that, just dropping it carelessly would do more harm than good.

However, drop-and-replace actually came out as a possible solution to a completely different sphere/topic we're dealing here:
_model/schema adjustments (on the go)_

See below.

**Comment by igr** [comment]
As long as I think about this, it becomes more and more important :) An update of the model is something that often happens with production code. So I am thinking about the following (similar to e.g. flybase):

+ there is additional storage (table) provided by the application that indicates application meta-data, like the model version.
+ if the current version != stored version (update of the app) that is an indication that we need to upgrade the model.
+ app will have upgrade methods: `(Dictionary) -> Dictionary` e.g. `v2_add_field`
+ what we need from Fridge is the following:
  1. add the ability to read data in a generic collection, like a map/dictionary
  2. add the ability to map the dictionary into the model (this can be done in 2 steps: dictionary -> json -> model)

In short, it would be great if we can have some kind of load/save method that uses a generic collection.

**Reply by vexy** [reply]
> An update of the model is something that often happens with production code.  

100% agree 👍 Often been in that situation myself.

I'm using `Fridge` on a live product, and yeah, I had a similar problem and similar approach to the problem as you've described.

I'll call it "_versioning strategy_".
Yup, model can be "versioned" somehow, ending up in some helper data structure that keeps up with _model variations_ over time.
But then, another question arises - what should you do when you figure a versioning mismatch ?

Say `struct A { var field1: Int }` now becomes `struct A { var field1: Int, var field2: Int }`. Okay `Fridge` will complain (throw error) but now what ? We need "expanded" or new `struct A`, however we already stored some data previously.

This is where "_migration step_" kicks in. As you've described above, there can be a helper table that keeps track of meta-structure.
Behind the scene, the real question is actually how to "migrate old structure to the new one". Say, there could be a mapping strategy that could traverse all fields of a _first_ struct and compare it to the _second_. It would be easy to compute differences and add them to the _second_ structure.
But consider the following (highly plausible, during development stages) scenario:

```swift
struct A {
  var field1: Int
}
```
Now we want it to become:
```swift
struct A {
   var field1: Int
   var field2: B
}

struct B {
  var field1: String
}
```

This is where "migration" starts to become a nightmare. 😄 

Again, handy approach would be to create new struct, `unfreeze` the old struct, compute the differences (during runtime) and `freeze` new struct back. This way, typically at the start of the program/app, some maintenance func takes over and compares the versions...

But, as this is everyday problem for all of us, this sounds interesting solution:
> add the ability to map the dictionary into the model (this can be done in 2 steps: dictionary -> json -> model)  

However, I'm just lacking to comprehend more complex schema changes.

Open to ideas and suggestions.

**Reply by igr** [reply]
That's what migration functions are for; created by a user. As long as Fridge can work with maps/dictionaries, the migration will work. Fridge doesn't care about migration. The user writes migrations manually for each version of data:

+ migrateFrom1To2()
+ migrateFrom2To3()
+ ...

Since input and output are generic containers, this will work. After migrating to the final state, Fridge saves the dictionary, and then the application continues to run. This will make the Fridge to load modified storage into the models of the latest version presented with the application.

**Reply by igr** [reply]
Something like `loadRaw` and `freezeRaw` methods;)

**Reply by igr** [reply]
I guess the following line would make be changed for the `raw` methods (from BSONConverter):

```
let realObject = try BSONDecoder().decode(T.self, from: rawBSONData)
```
i.e. this one in the Fridger:
```
let storedObject: T = try reader.read()
```
Instead of T in the 'raw' methods we might use Dictionary[String, Any]? Thinking out loud... its nothing urgent :)

**Reply by vexy** [reply]
Converting concrete type to `Any` would lead to _type erasure_ problem and would lose all the benefits of generic approach 🫤

**Reply by igr** [reply]
Yeah, I was thinking of a separate method. Anyhow

```let a: Dictionary<String, String> = try Fridge.unfreeze🪅🎉(objId)```

does not work, even I have all the strings inside.

Nevermind! I just wanted to avoid carrying around all the model versions, but that might not be a that bad idea after all!

**Reply by igr** [reply]
We can close this question now :) Btw its not a Bug
